### ISSUE TITLE
[$1000] Android - Copy / Paste / Cut menu is not displayed when selecting text in compose box - Reported by: @parasharrajat 

### ISSUE THREAD

**Issue by isagoico** [Expensify/App#6876]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Navigate to a conversation
2. Type anything in compose box
3. Select it 
## Expected Result:
Native menu to cut / copy / paste should be displayed 

## Actual Result:
Native menu to cut / copy / paste is displayed for a second and then it's hidden

## Workaround:
User is unable to cut / copy / paste text in the compose box. 
## Platform:

Where is this issue occurring?

- Android

**Version Number:** 1.1.23-0

**Reproducible in staging?:** yes 
**Reproducible in production?:** yes 

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/44479856/147117514-155c964a-587d-4581-82d6-5b842041902a.mp4

**Expensify/Expensify Issue URL:**

**Issue reported by:** @parasharrajat 
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1640179428463500

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @neil-marcellini (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @Christinadobrzyn (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by Christinadobrzyn** [comment]
Upwork job created!

Internal: https://www.upwork.com/ab/applicants/1473877014903103488/job-details
External: https://www.upwork.com/jobs/~01333b75dd01dc96cb

Hired @parasharrajat for reporting this. Waiting on proposals for the fix.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @parasharrajat (`Exported`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @stitesExpensify (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by kursat** [comment]
React native's TextInput re-renders when its selection property is controlled. There is an open issue (https://github.com/facebook/react-native/issues/29063) about this.

I'm not sure how convenient is this but we did it for ios, so I'm proposing to omit selection prop at `TextInputFocusable/index.android.js`.

```js
    render() {
        const propsToPass = _.omit(this.props, 'selection');

        return (
            <TextInput
                placeholderTextColor={themeColors.placeholderText}
                ref={el => this.textInput = el}
                maxHeight={CONST.COMPOSER_MAX_HEIGHT}
                rejectResponderTermination={false}
                editable={!this.props.isDisabled}
                /* eslint-disable-next-line react/jsx-props-no-spreading */
                {...propsToPass}
            />
        );
    }
```

**Comment by parasharrajat** [comment]
Will it affect our app's functionality? @kursat

**Comment by kursat** [comment]
@parasharrajat, sorry to say but yes it will. At first glance I noticed this: 

When we replace a selected text  with an emoji, we will not be able to move cursor at the end of emoji
https://github.com/Expensify/App/blob/d6a607324f84e654da2da6e073b150a00def7490/src/pages/home/report/ReportActionCompose.js#L418-L433

I tried to replace it like below, but setting selection to a js object like this, fixes position of cursor and you literally write backwards after than that. 
```js
        this.textInput.setNativeProps({
            selection: {
                start: this.state.selection.start + emoji.length,
                end: this.state.selection.start + emoji.length,
            },
        });
```

We are also passing selection prop from a couple files, but for now I am not sure about the impact yet. I'll work on this one more day and see if I can eliminate all possible side effects.

**Comment by parasharrajat** [comment]
Any update @kursat

**Comment by kursat** [comment]
Hey @parasharrajat, I was dealing some personal problems, I totally forgot about this. So sorry about the absurd delay.

The only component that sets selection is ReportActionCompose. To keep selection setting feature we can manage selection state in `TextInputFocusable` component. And by setting another ref for `TextInputFocusable` itself, we can set selection state programmatically.

src/components/TextInputFocusable/index.android.js:

```js
class TextInputFocusable extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            selection: null,
        };
    }

    // ...

    setSelection = (selection) => {
        this.setState({
            selection,
        });
    }

    render() {
        const propsToPass = _.omit(this.props, 'selection');
        return (
            <TextInput
                placeholderTextColor={themeColors.placeholderText}
                ref={el => this.textInput = el}
                maxHeight={CONST.COMPOSER_MAX_HEIGHT}
                rejectResponderTermination={false}
                editable={!this.props.isDisabled}
                selection={this.state.selection}
                /* eslint-disable-next-line react/jsx-props-no-spreading */
                {...propsToPass}
            />
        );
    }
}

// ...

export default React.forwardRef((props, ref) => (
    /* eslint-disable-next-line react/jsx-props-no-spreading */
    <TextInputFocusable {...props} ref={props.ownRef} forwardedRef={ref} />
));
```

src/pages/home/report/ReportActionCompose.js:

```js

    constructor(props) {
        super(props);

        // ...
        this.setTextInputFocusableRef = this.setTextInputFocusableRef.bind(this);

    }

    setTextInputFocusableRef(el) {
        this.textInputFocusable = el;
    }

    addEmojiToTextBox(emoji, emojiObject) {

        // ...

        if (this.textInputFocusable) {
            this.textInputFocusable.setSelection({
                start: this.state.selection.start + emoji.length,
                end: this.state.selection.start + emoji.length,
            });
        }

        // ...
    }

    render() {
        <TextInputFocusable
          autoFocus={this.shouldFocusInputOnScreenFocus || _.size(this.props.reportActions) === 1}
          multiline
          ref={this.setTextInputRef}
          ownRef={this.setTextInputFocusableRef}
          { /* ... */ }
        />
        // ...
    }
}
```

This way we don't brake anything.

**Comment by parasharrajat** [comment]
Great thanks. I see. If we are able to migrate from selection prop to `this.textInput.setSelection` API then I would love that solution. But I need to look at the code to analyze it before we proceed. It is going to be a big change. Better to involve @marcaaron. 
we are looking to make these fully uncontrolled and this could be a good initiative.

**Comment by kursat** [comment]
Happy to help! You can experiment on this fork if it helps: https://github.com/kursat/App/tree/proposal-6876

**Comment by marcaaron** [comment]
Would like to clarify some things about this proposal to be sure I understand. 

Is this a good description of the problem and solution?

- `TextInput` won't re-render when the `selection` prop changes
- So we should force it to re-render by manually setting the selection state

If so, this seems like a workaround and ideally, we would get it fixed at the `react-native` level or at the very least open a tracking issue to see if we can unwind the proposed changes when it is fixed upstream.

But also wondering if we can move the logic so that a re-render is forced when the props change as a side effect instead of programmatically triggering them with `textInputFocusableRef.setSelection()`. It seems this would be easier to remember and require less manual setting of the selection?

Would something like this work in `TextInputFocusable/index.android.js`?

```js
componentDidUpdate(prevProps) {
    if (_.isEqual(prevProps.selection, this.props.selection)) {
        return;
    }

    this.setState({selection: this.props.selection});
}
```

**Comment by parasharrajat** [comment]
I was thinking that it could be a good opportunity for us to drop the selection prop which makes it controlled. We only require setting the selection when we want to place the emoji so if we can do it directly via API then I think that is better.

**Comment by marcaaron** [comment]
>We only require setting the selection when we want to place the emoji so if we can do it directly via API then I think that is better.

Got it. Just to make sure I understand the point of view, why this is better?

>I have seen a few side effects. Sometimes cursor jump between positions when we move the cursor placement manually on Android. I believe its the selection prop.

Are there bug reports for these with reproductions? Is there a known connection between passing a `selection` prop and seeing the "cursor jump"?

**Comment by parasharrajat** [comment]
> why is this better?

Using selection prop will make it controlled and if use API  it will remain uncontrolled. I think that would be better performance-wise.

> Are there bug reports for these with reproductions?

I haven't reported it yet. I will do that.

> Is there a known connection between passing a selection prop and seeing the "cursor jump"?

It is just a hunch. I haven't debugged this yet.  Due to the jump that happens when we move the cursor and cursor positioning is directly related to selection, I believe that is the cause. I will report it and then we can get more eyes on it.

**Comment by marcaaron** [comment]
>Using selection prop will make it controlled and if use API it will remain uncontrolled. I think that would be better performance-wise.

Oh right, because the same can be observed with a controlled `value` prop? That could be! I haven't observed any performance issues with the `selection` prop but it can probably be tested before making a decision. Using a controlled `value` definitely feels slow, is the same true for `selection`? Maybe we can test and figure that out as part of this issue.

**Comment by Christinadobrzyn** [comment]
Raised the [Upwork job](https://www.upwork.com/ab/applicants/1473877014903103488/job-details) price to $500. It looks like we're still reviewing proposals correct @parasharrajat?

**Comment by parasharrajat** [comment]
We have a proposal but I need to rigorously review it before we can proceed. @Christinadobrzyn

**Comment by parasharrajat** [comment]
@kursat I checked your proposal. It has an issue. Sometimes, when we manually move the cursor to someplace and then add another emoji. Emoji is added back to the previous position and the cursor will remain there even if we try to drag it to another location. 

Could you please prepare a final proposal with all the steps? Also, test the cursor placement carefully. 

And What do you think about https://github.com/Expensify/App/issues/6876#issuecomment-1005176447?

**Comment by kursat** [comment]
Apparently keeping selection variable in the state still not solving re-render issue. It looked like it did but I must have missed something while I am testing. I need to start over, for now please ignore my proposal.

**Comment by Christinadobrzyn** [comment]
Increased job price to $1000

**Comment by K4tsuki** [comment]
#### proposal

The problem is setting `selection` property of `TextInput` using state value of `selection` and set this value of text  input selection in `onSelectionChange` could be triggering text input re-render before context menu displayed. 

https://github.com/Expensify/App/blob/0bce09177d2aed8c5e377bc758f5241dd22886eb/src/pages/home/report/ReportActionCompose.js#L207-L209

https://github.com/Expensify/App/blob/0bce09177d2aed8c5e377bc758f5241dd22886eb/src/pages/home/report/ReportActionCompose.js#L650-L651

Selection value is used in `addEmojiToTextBox`.

My solution is to remove  selection property in `ReportActionCompose` and set selection value only after adding emoji and move selection property to `TextInputFocusable` in android file:
in `TextInputFocusable` we add condition in render  file whether we should use `selection` or not. This could be:

<details> <summary> Code: </summary>

```
render() {
        console.log('focusable render');
        if ( this.isSelectionChanged ) {
            this.isSelectionChanged = false;
            return ( 
                <TextInput
                    placeholderTextColor={themeColors.placeholderText}
                    {...this.props}
                    ref={el => this.textInput = el}
                    maxHeight={CONST.COMPOSER_MAX_HEIGHT}
                    rejectResponderTermination={false}
                    editable={!this.props.isDisabled}
 +                 selection={this.selectionvalue}
                    /* eslint-disable-next-line react/jsx-props-no-spreading */
                />);
        } else {
            return ( 
                <TextInput
                    placeholderTextColor={themeColors.placeholderText}
                    {...this.props}
                
                    ref={el => this.textInput = el}
                    maxHeight={CONST.COMPOSER_MAX_HEIGHT}
                    rejectResponderTermination={false}
                    editable={!this.props.isDisabled}
                />);
        } 
    }

```
</details>

And we `createRef` element for `TextInputFocusable` by `this.props.forwardedRef(this.textInput, this);` in  :

https://github.com/Expensify/App/blob/0bce09177d2aed8c5e377bc758f5241dd22886eb/src/components/TextInputFocusable/index.android.js#L53

and in`setTextInputRef` add second parameter `focusElement`, this could be:
```
setTextInputRef(el, textInputFocusableEl) {
        ReportActionComposeFocusManager.composerRef.current = el;
        
        if ( textInputFocusableEl ) { this.textInputFocusable = textInputFocusableEl; }
        this.textInput = el;
    }

```
`

Then provide `setSelection` function inside `TextInputFocusable` to set variables `isSelectionChanged` and `selectionvalue` inside render function.

In `ReportActionCompose`, save selection value inside object property and `onSelectionChange` could be: `this.selectionValue =  e.nativeEvent.selection`.

Then in `addEmojiToTextBox`

We call `this.focusableInput.setSelection` using saved `this.selectionValue` before `updateComment` and after `updateComment`, this could be:

<details> <summary> Code : </summary>

```
if ( this.focusableTextInput && this.focusableITextnput.setSelection ){
    this.focusableTextInput.setSelection( this.selectionValue );
}

this.updateComment(newComment);

if ( this.focusableTextInput && this.focusableTextInput.setSelection ){
     setTimeout( () => {
         const pos = this.selectionValue.start + emoji.length;
         this.focusableTextInput.setSelection( { start: pos, end: pos} );
     }, 10);
}
```

</details>

`setTimeout` here is necessary to wait for textInput being updated, even small number `10` could be used here.


<details> <summary> Result: </summary>


https://user-images.githubusercontent.com/1313124/150620842-8eb6e96f-2905-4bce-a6dc-6f3c7dd1d30d.mp4



</details>

**Comment by parasharrajat** [comment]
Thanks for the proposal @K4tsuki. There are few things to consider before we move forward in finding a solution to this problem.

> setTimeout here is necessary to wait for textInput being updated, even small number 10 could be used here.
That is my solution or something similar along the line.

Our codebase strictly prohibits the usage of setTimeout unless this is the last resort and we don't have control over the API. So we should try to get away from it as much as possible. 

> Code: 

In your first code example, you are remounting the TextInput which is not really the React way. The same thing should be achieved either from props or the state.

Let's try to find alternatives to these.

**Comment by K4tsuki** [comment]
Update I could solve the problem by adding parameter to `updateNewComment` with new `nextSelection` parameter that will be used by `addEmojiToBox`: 

```
if ( nextSelection ) { 
            this.textInput.setNativeProps({text: newComment   , selection: nextSel });
        } else {
            this.textInput.setNativeProps({text: newComment   , selection: {start: undefined, end: undefined} });
        }

```
in :

https://github.com/Expensify/App/blob/07582967349266b430103b11a94839bf40d8bccb/src/pages/home/report/ReportActionCompose.js#L336

more detail maybe later.

**Comment by K4tsuki** [comment]
#### proposal

I am able to find simpler solution:
Remove selection property in 

https://github.com/Expensify/App/blob/07582967349266b430103b11a94839bf40d8bccb/src/pages/home/report/ReportActionCompose.js#L664

<br />

Then In `updateComment` 

https://github.com/Expensify/App/blob/07582967349266b430103b11a94839bf40d8bccb/src/pages/home/report/ReportActionCompose.js#L335-L336

add new parameter `nextSelection`,  and change that line to be:

```
function updateComment(newComment, nextSelection) {
  const selection = nextSelection
    ? nextSelection
    : { start: undefined, end: undefined };
  this.textInput.setNativeProps({ text: newComment, selection });
}
```

using `{ start: undefined, end: undefined }` is like releasing selection value to be native handled.

<br />

Then `onSelectionChange` could be:
```
if (!this.state.isEmojiPickerVisible) {
  this.currentSelection = e.nativeEvent.selection;
}
```
no need to save `selection` in `state`.
Also `isEmojiPickerVisible` check is used because when emoji is shown, selection is changed, this could make wrong position or selection when adding emoji.
Also check on `state.isEmojiPickerVisible` maybe will be late. using simple variable `this.isEmojiPickerShown` above `blur` and `setState` in `showEmojiPicker`: will be more reliable.

https://github.com/Expensify/App/blob/07582967349266b430103b11a94839bf40d8bccb/src/pages/home/report/ReportActionCompose.js#L407-L408


Then `addEmojiToTextBox` will be:
<details> <summary> addEmojiToTextBox  </summary>

```
function addEmojiToTextBox(emoji, emojiObject) {
  EmojiUtils.addToFrequentlyUsedEmojis(
    this.props.frequentlyUsedEmojis,
    emojiObject
  );
  this.hideEmojiPicker();

  const newComment =
    this.comment.slice(0, this.currentSelection.start) +
    emoji +
    this.comment.slice(this.currentSelection.end, this.comment.length);

  const nextSelection = {
    start: this.currentSelection.start + emoji.length,
    end: this.currentSelection.start + emoji.length,
  };

  this.updateComment(newComment, nextSelection);
}

```

</details>
There is no need to set `textInput.setNativeProps({ text: newComment, });` because `updateComment` set that property too so I am removing it.

**Comment by parasharrajat** [comment]
Great. Thanks for the updates. I will review this shortly.

**Comment by parasharrajat** [comment]
>Then onSelectionChange could be:

> Also isEmojiPickerVisible check is used because when emoji is shown, selection is changed, this could make wrong position or selection when adding emoji.
> Also check on state.isEmojiPickerVisible maybe will be late. using simple variable this.isEmojiPickerShown above blur and setState in showEmojiPicker: will be more reliable.

I am trying to digest this information. 

- When you set a value via `setnativeProps` API It should not trigger a selection update. Right? 
- Why the state.isEmojiPickerVisible will be late? Maybe it is more reliable than normal instance property.

> There is no need to set textInput.setNativeProps({ text: newComment, }); because updateComment set that property too so I am removing it.

There was one old bug (can't remember the link) that added this. You can find this info via git blame.

**Comment by K4tsuki** [comment]
> When you set a value via setnativeProps API It should not trigger a selection update. Right?
No, if you not specify the `selection` field.

When user open `emojiPicker` `textInput` get blurred and selection change. And this is will change user selection right before he open `emojiPicker` . I mean in this code:

https://github.com/Expensify/App/blob/a9dd57ca3f09181df92699e7c10b6cd7dbddfaaa/src/pages/home/report/ReportActionCompose.js#L250-L253

On `this.textInput.blur();` the selection will change and will be different than what user has set the selection. So setting variable like `this.isEmojiPickerShown` to true before blur, to signal  `don't save selection value at the moment` to prevent selection change when getting blurred. Then change `isEmojiPickerShown` back to false when hide `emojiPicker`.


<br />

> There was one old bug (can't remember the link) that added this. You can find this info via git blame. 

https://github.com/Expensify/App/commit/a9dd57ca3f09181df92699e7c10b6cd7dbddfaaa

`textInput.setNativeProps({ text: newComment, });` First appears in there. If you look at the file `ReportActionCompose` on that version `updateComment` doesn't had `textInput.setNativeProps({ text: newComment, });` at that time. 

Currently there is available `textInput.setNativeProps({ text: newComment, });` in `updateComment` so removing it wouldn't be a problem.

**Comment by parasharrajat** [comment]
> When user open emojiPicker textInput get blurred and selection change.

I think when input is blurred no selection event is fired.

**Comment by K4tsuki** [comment]
> I think when input is blurred no selection event is fired.

I  am not sure but just to be safe. I have tested it using `isEmojiPickerVisible` state, but  somehow it is late and selection get changed using development environment. So just to be really safe here we use another variable `isEmojiPickerShown`. 
If you don't really like it, using state `isEmojiPickerVisible` is fine.

**Comment by parasharrajat** [comment]
>  I have tested it using isEmojiPickerVisible state, but somehow it is late and selection get changed using development environment.

Now we are talking about the real problem. In the previous proposals, I faced the same issue and it seems you are facing the same. You can play with it https://github.com/Expensify/App/issues/6876#issuecomment-1004923865. 

Let's try to find the real cause of it. This is not part of the issue but when we will adopt this solution, this will become an issue for us.

**Comment by K4tsuki** [comment]
What issue actually, it is only  used in to filter out `onSelectionChange` :

```
onSelectionChange() {
   if (this.isEmojiPickerShown) {
      this.currentSelection = e.nativeEvent.selection;
    }
}

```

It is only used here I think the problem is different from that comment. 
This is only used when emoji picker is shown and not all the time to filter out and signaling `we don't want to save selection change after emoji picker button is pressed` or something like that.

**Comment by parasharrajat** [comment]
For the reason that you explained for this change is not based on any fact thus it makes it unpredictable. Also, I don't think this code is needed. You are saying that this code fixes some issue which makes it a hack.

**Comment by K4tsuki** [comment]
Without that code emoji placement is wrong, so it fixes the issue. I recommend you try the code to see it yourself.

**Comment by parasharrajat** [comment]
I am not rejecting the proposal but asking to find a future proof and alternative.

**Comment by K4tsuki** [comment]
>  future proof 
is to to set `this.isEmojiPickerShown = true;` in the `showEmojiPicker`  at first line.
and set `this.isEmojiPickerShown = false;` in `hideEmojiPicker`

**Comment by K4tsuki** [comment]
@marcaaron could you review my [proposal](https://github.com/Expensify/App/issues/6876#issuecomment-1021918193) please, I really appreciate it.

**Comment by parasharrajat** [comment]
Great to have a second opinion but until I see a clear reason for https://github.com/Expensify/App/issues/6876#issuecomment-1022773821. I can't move forward with this. 

>  I recommend you try the code to see it yourself.

I will try your changes as well. Overall I see that your solution works, it's just every piece of code should have a clear reason behind it.

**Comment by K4tsuki** [comment]
> clear reason for #6876 (comment):

This is the reason: selection is changed when user open emoji picker. 
I am just adding 1 variable here, and this is just proposal. Why really bothered I don't understand. 
And also the code doesn't disturb other code.

**Comment by parasharrajat** [comment]
> selection is changed when user open emoji picker.

I didn't observe this. and this should not happen because the whole logic of adding emojis depends on this fact. If I am not wrong your code will prevent change in selection when it injects emoji in the text and then you are setting the position manually. 

While emojiPicker is open, there is no way that selection will change as textinput is not focused at all during that time.

But let's say the selection was changed when you added emoji and you again programmatically set the cursor position. Shouldn't both be the same? I think they should not conflict.

**Comment by K4tsuki** [comment]
> your code will prevent change in selection when it injects emoji in the text and then you are setting the position manually.

correct. I am setting the position to last set by user right before opening emoji picker.

> While emojiPicker is open, there is no way that selection will change as textinput is not focused at all during that time.

Correct. The selection is not changed  while emojiPicker is opens. but at the start when emojiPicker button is pressed to show the emoji picker.

> But let's say the selection was changed when you added emoji and you again programmatically set the cursor position. 

No, while selecting scrolling or viewing or chose an emoji. the selection is not changed because `textInput` already blurred.
The change is right at the start, of `showEmojiPicker`.  I think when `this.textInput.blur();`

**Comment by marcaaron** [comment]
@K4tsuki Sorry, I'm not assigned to this issue
@parasharrajat can you review the proposal and @stitesExpensify will sign off if it looks good

**Comment by K4tsuki** [comment]
All right @marcaaron, I understand.

**Comment by parasharrajat** [comment]
@K4tsuki's [proposal](https://github.com/Expensify/App/issues/6876#issuecomment-1021918193) here looks good but I didn't get a good sense of why https://github.com/Expensify/App/issues/6876#issuecomment-1022773821 is used. It does not make sense based on how the selection changes. I tried to get the proper explanation in the above comments. But it seems that it is necessary for @K4tsuki 's proposal to work.  So I am giving it a :green_circle: signal. 

cc: @stitesExpensify 

:ribbon: :eyes: :ribbon:  C+ reviewed

**Comment by Christinadobrzyn** [comment]
Just confirming @parasharrajat - are we good to hire @K4tsuki for this fix?

**Comment by parasharrajat** [comment]
Not yet @Christinadobrzyn. @stitesExpensify will review the proposal first.

**Comment by stitesExpensify** [comment]
I think we can move forward with @K4tsuki proposal. @parasharrajat I appreciate your concerns and will keep my eyes on them when the PR goes up, but as of right now I can't see definitely that something will go wrong so I'm okay with giving it a shot!

**Comment by Christinadobrzyn** [comment]
Sounds good! @parasharrajat is already hired as the Contirbutor+ for this issue. He'll also get the $250 bonus for reporting this issue. 

@K4tsuki what is your name in Upwork so I can hire you?

**Comment by K4tsuki** [comment]
@Christinadobrzyn here is my upwork profile:
https://www.upwork.com/freelancers/~01bdd9e3a6f15042bd

**Comment by Christinadobrzyn** [comment]
Thanks @K4tsuki - sent you an offer! Thanks!

**Comment by stitesExpensify** [comment]
PR is up but there is an issue with iOS

**Comment by JalalMitali** [comment]
### Proposal : 
In TextInputFocusable/index.android.js I added the style : 
`{height: Platform.OS == 'android' ? 40 : 20}`
and it currently seems to be working :)

**Comment by JalalMitali** [comment]
### Proposal : 
1. add a state variable selection initialized with start set to 0 and end set to 0 instead of comment
2.`onSelectionChange={(e) => this.setState({selection: e.nativeEvent.selection})}` in index.android.js instead of Report Action Compose.
3. comment the selection in Report Action Compose.
4. that is it! 
5. I am not sure if this is what you need or there is another use for comment you may correct me if I am wrong :(

**Comment by kidroca** [comment]
This issue was brought for feedback on slack: https://expensify.slack.com/archives/C01GTK53T8Q/p1645191589301989 and since it's accepting proposals again (_help wanted_ label) here's mine: 

#### Problem:
Controlled selection (a selection that is set and updated on props) is getting in the way of React to update the selection from user interactions

#### Solution: Switch to uncontrolled selection 
Switching to uncontrolled selection addresses the issue and improves performance (less re-renders):
1. Introduce a `defaultSelection` prop to `TextFieldFocusable`. Similarly to how [defaultValue](https://reactnative.dev/docs/textinput#defaultvalue) works for values - it would set selection for the first render of the field and then unset it after mount
2. Components that need information about current selection `ReportActionCompose` can track that outside of `state` (unless they really need to cause a re-render when it changes):
    ```js
    onSelectionChange(e) {
      this.currentSelection = e.nativeEvent.selection;
    }
    ```
3. These changes are enough to turn `addEmojiToTextBox` to this:
    ```js
    addEmojiToTextBox(emoji) {
      const newComment = this.comment.slice(0, this.currentSelection.start)
        + emoji + this.comment.slice(this.currentSelection.end, this.comment.length);

      this.updateComment(newComment);
    }
    ```
   Now the copy/paste context menu appears for selection and inserting emojis still works as before

https://user-images.githubusercontent.com/12156624/155128799-498584ca-1634-4e91-a51c-6b7f3dbe1b48.mp4

My minimal changes PR from the slack thread is here: https://github.com/Expensify/App/pull/7815

---

We can go up one level and include the following updates as well 

1. Abstract `setNativeProps` usages
    ```js
      this.textInput.setNativeProps({text: newComment});
    ```
    becomes:
    ```js
      this.textInput.updateText(newComment);
    ```
2. Abstract `addEmojiToTextBox` usage
    Introduce a `insertText` method and update usage like
    ```js
      addEmojiToTextBox(emoji) {
         const newComment = this.textField.insertText(emoji, this.currentSelection);
         this.updateComment(newComment);
      }
    ```
Now if something platform specific should happen during `updateText|insertText` it can happen inside `TextInputFocusable/index.(ios|web|android).js`

**Comment by parasharrajat** [comment]
Sorry guys for the wrong impression. @K4tsuki is already hired for this job and PR is in review.

**Comment by parasharrajat** [comment]
No update on slack https://expensify.slack.com/archives/C01GTK53T8Q/p1645191589301989. Pinged again.

**Comment by parasharrajat** [comment]
Thanks for your patience @K4tsuki. [stitesExpensify](https://github.com/stitesExpensify) is OOO. Let's wait for him. 

Update: Discussion happening on the PR...

**Comment by stitesExpensify** [comment]
Discussion still on PR

**Comment by stitesExpensify** [comment]
Discussion still on PR

**Comment by mallenexpensify** [comment]
The previous hire is not working on this issue any more, they've been compensated for their time in Upwork. @stitesExpensify or @parasharrajat would you prefer to fix the issue based on the work that's already been done in the PR or should we reopen and hire another contributor to implement?

**Comment by parasharrajat** [comment]
I am still not :100: up with the PR changes. I would prefer a new and better implementation. But currently, I don't have any idea of the solution.

**Comment by stitesExpensify** [comment]
I agree, I think hiring another contributor is the best way forward

**Comment by mallenexpensify** [comment]
Added `Help Wanted` and `Exported`, created a new job in Upwork https://www.upwork.com/jobs/~013607bfacea338b43

**Comment by kidroca** [comment]
I've submitted a proposal some time ago, but a contributor was already working on the ticket, maybe you can consider it now: https://github.com/Expensify/App/issues/6876#issuecomment-1047760646

The PR from the slack thread: https://github.com/Expensify/App/pull/7815

**Comment by mallenexpensify** [comment]
@parasharrajat , can you review @kidroca 's proposal and PR above?

**Comment by parasharrajat** [comment]
I will @mallenexpensify. Reviewing others ATM.

**Comment by parasharrajat** [comment]
@kidroca has a clean start but there are a few problems that we faced during implementation. 
1. Selection works differently on IOS|ANDROID|WEB. Can you confirm if your proposal works on all platforms? I don't see any steps to solve those in the proposal. 

Also, PR is outdated.

**Comment by kidroca** [comment]
> 1. Selection works differently on IOS|ANDROID|WEB. Can you confirm if your proposal works on all platforms

Yes, the change solves the issue on all platforms

Making `selection` uncontrolled fixes the issue for android/ios

This change is covering web/desktop:

```
        const caret = this.currentSelection.start + emoji.length;

        if (this.textInput.setSelectionRange) {
            this.textInput.setSelectionRange(caret, caret);
        }
```

We can move the web/desktop specific change here: https://github.com/Expensify/App/blob/main/src/components/TextInput/index.js

Instead of passing a `defaultSelection` that never changes (to achieve the uncontrolled mode) we can introduce a `defaultSelection` prop on our `TextInput` components that

Everything depends on what we want to cover 
- if all we want to fix is the ReportActionCompose the changes on the PR cover all platforms 
- if we want to support uncontrolled `selection` on all `TextInput` components we can implement a `defaultSelection` prop

**Comment by kidroca** [comment]
Sorry, I guess I've only tested popping the context menu on ios and not inserting an emoji.

Tested it again and added a fix to preserve caret position after inserting an emoji, by using the interaction manager
- PR updated with 4 more line changes: https://github.com/Expensify/App/pull/7815

```js
    addEmojiToTextBox(emoji) {
        const newComment = this.comment.slice(0, this.currentSelection.start)
            + emoji + this.comment.slice(this.currentSelection.end, this.comment.length);

        this.updateComment(newComment);

        const caret = this.currentSelection.start + emoji.length;

        InteractionManager.runAfterInteractions(() => {
            this.textInput.setNativeProps({selection: {start: caret, end: caret}});
        });

        if (this.textInput.setSelectionRange) {
            this.textInput.setSelectionRange(caret, caret);
        }
    }
```

https://user-images.githubusercontent.com/12156624/159985807-71d3f8a9-fa82-4479-8a38-f8b5100b8805.mov

---

Seeing that we have access to `this.textInput.selectionStart` we might not need to capture `this.currentSelection` at all

**Comment by parasharrajat** [comment]
Did you test it on Android as well? I think it will have some problems there too. in the previous PR we have to loose the selection after immediately setting it.

**Comment by kidroca** [comment]
Yeah there's an issue on Android:

Here's a summary:

1. If we set `selection` prop, context menu does not work on Android
2. If we don't set native props, like originally suggested, everything works kind-a OK, with the exception that if you select text and then replace it with an emoji, then the emoji remains selected 
3. If we set selection through native props the carret starts to act weird always jumping to the position we set by `setNativeProps`, so we have to quickly clear what we just set

```js
InteractionManager.runAfterInteractions(() => {
  this.textInput.setNativeProps({selection: {start: caret, end: caret}});
    global.requestAnimationFrame(() => this.textInput.setNativeProps({selection: null}));
  });
```

But setting `{ selection: null }` messes it for iOS where, the behavior is the same as 2) above - after inserting an emoji it's selected

---

In light of this, Seeing that we have `Composer/index.(ios|android).js`, I would suggest to introduce a `setSelection` method and have different implementations based on platform:

Web/Desktop:
```js
setSelection(start, end) {
  this.textInput.setSelectionRange(start, end);
}
```

iOS:
```js
setSelection(start, end) {
        InteractionManager.runAfterInteractions(() => {
            this.textInput.setNativeProps({selection: {start, end});
        });
}
```

Android:
```js
setSelection(start, end) {
        InteractionManager.runAfterInteractions(() => {
            this.textInput.setNativeProps({selection: {start, end});
            global.requestAnimationFrame(() => this.textInput.setNativeProps({selection: null}));
        });
}
```

And in `ReportActionCompose`:

```js
    addEmojiToTextBox(emoji) {
        const newComment = this.comment.slice(0, this.currentSelection.start)
            + emoji + this.comment.slice(this.currentSelection.end, this.comment.length);

        const caret = this.currentSelection.start + emoji.length;
        this.updateComment(newComment);
        this.textInput.setSelection(caret, caret);
    }
```

**Comment by kidroca** [comment]
Alternatively we can find/submit a bug report regarding Android, where using `setNativeProps` causes point 3) above, and we have to use `setNativeProps` again to clear the `selection`

**Comment by parasharrajat** [comment]
@stitesExpensify so we kind of ended with the same solution as in the previous PR. I think are a few bugs in RN. But RN suggests using controlled props and they prevent cursor issues but on Android, the menu is not opening. 

so my suggestion would be we just file a report on RN for the native Copy-paste menu and keep the same composer implementation. 

What do you say?

**Comment by stitesExpensify** [comment]
Yep, a fix to RN is always preferable over a workaround in our codebase. We can file a report, but also I believe that if a contributor creates a PR to fix RN we still pay out the upwork issue even though it's not in our codebase. Is that correct @mallenexpensify ?

**Comment by kidroca** [comment]
Controlled `selection` is defeating the purpose of the uncontrolled TextInput, it causes a re-render each time we input a character - something we originally tried to avoid

The reason we're not setting `value` on the TextInput is to avoid re-rendering for each keystroke
I guess we've missed this when `selection` was added to the component

https://user-images.githubusercontent.com/12156624/160128418-e3193dd6-813f-47d2-ad1c-dbadcd6857f8.mp4

_You can also observe how words sometime get selected while you type - that's another side effect of controlled selection_

A problem with controlled props in general is that a component doesn't know we're echoing back it's own value. And echoing can be delayed resulting in bugs like selecting previous words while you type
Specifically for react-native echoing results in sending data back and forth through the native bridge, and this bridge is the bottleneck of react-native

---

It's seems, on Android, setting `selection` through props or through `setNativeProps` doesn't work as expected, the cause might be the same, but ideally the compose box would be more performant if it sets `selection` only when necessary through `setNativeProps`. Perhaps the focus for the react-native ticket should be on `setNativeProps` and not on controlled `selection`

**Comment by kidroca** [comment]
> Yep, a fix to RN is always preferable over a workaround in our codebase.

How do we decide that?
For example we have this code, clearly stating it's a workaround:

https://github.com/Expensify/App/blob/c04fca7d9bd69ee8717f81714b41c0be60acde38/src/pages/home/report/ReportActionCompose.js#L272-L276

What I'm suggesting is similar, it also uses the `InteractionManager` and delays code using `requestAnimationFrame` (like `setTimeout`), yet my proposal is considered a "workaround better to be addressed in RN"

What I'm proposing is a fix of about 30 line changes, that also addresses the re-render per keystroke issue vs finding/waiting for someone to identify the issue in Android and submit a fix to react-native 

We can apply my fix, and all we have to do would be to remove one line - `requestAnimationFrame`, **when** the issue in Android is resolved

**Comment by parasharrajat** [comment]
I have noticed that controlled props perform more badly on dev builds than PROD. I think ReactActionComposer is also lacking optimizations. A single render takes a bit of time. If optimize this. We can get better results. 

it has many things to render
1. AttachmentModal
2. AttachmentPicker.
3. PopoverMenus.
4. TextInput. 
5. Modals. 

Also, inline Children is making it worse. https://github.com/Expensify/App/blob/9c50db754bc29b340f7f725059d0cca6771fc61f/src/pages/home/report/ReportActionCompose.js#L425

**Comment by stitesExpensify** [comment]
>> Yep, a fix to RN is always preferable over a workaround in our codebase.

> How do we decide that?

I'm taking that from our design philosophy [here](https://github.com/Expensify/App#philosophy) Section 5.2

> For example we have this code, clearly stating it's a workaround:

I think the main difference is that the maintainer of that package is not willing to update it whereas in this case the problem is that the platforms are not in sync because of react native itself. I think that in general, we should avoid workarounds. 

IMO any solution that involves a timeout or similar mechanism is not ideal, including the code that you linked.

**Comment by rushatgabhane** [comment]
@sig5 you mind weighing in here? Your work on 
 IOUAmount #7433  should help you solve this issue too.

**Comment by kidroca** [comment]
Thanks for the answers @stitesExpensify, sounds fair 

As far as I see there's no way to address this issue on the javascript side then. We should make that clear when summoning other people to solve the problem, or at least let them know what have been tried so far, so they don't waste as much time ending on the same crossroad

Expensify has a fork of `react-native` where contributors can submit a PR with a fix for Android
Then Expensify submits a PR upstream and uses the forked version in App in the meantime

**Comment by stitesExpensify** [comment]
Great, sounds good. Do you think that we should keep this issue open, or create a new one specifying the solution should be in the fork?

**Comment by kidroca** [comment]
I lean toward creating a new ticket and summarizing what have been found out here

**Comment by mallenexpensify** [comment]
Missed this last week
>Yep, a fix to RN is always preferable over a workaround in our codebase. We can file a report, but also I believe that if a contributor creates a PR to fix RN we still pay out the upwork issue even though it's not in our codebase. Is that correct @mallenexpensify ?

Yes, if someone fixes a bug anywhere that helps with the issue, they'll be compensated.

If we close this issue and open another, lets make sure to keep @parasharrajat on as C+.
@K4tsuki has already been compensated so we can close their PR (and anyone can use any of the code from there if wanted since @K4tsuki has been paid in full)

**Comment by mallenexpensify** [comment]
@kidroca can you create the new issue, count the time as part of your hourly work, include a link back to this issue and cc all active participants in this convo in the OP of the new issue? (I'll assign myself as CM and others once done)

**Comment by kidroca** [comment]
@mallenexpensify

> @kidroca can you create the new issue, count the time as part of your hourly work, include a link back to this issue and cc all active participants in this convo in the OP of the new issue? (I'll assign myself as CM and others once done)

Yes, I'll post back when ready

**Comment by kidroca** [comment]
Created a new ticket of the same name here: https://github.com/Expensify/App/issues/8349

**Comment by mallenexpensify** [comment]
Thanks @kidroca , should we close this one now?
I think we should double the price of the other, anyone object? cc @NicMendonca

**Comment by mallenexpensify** [comment]
Closing, doubled price of the[ other one](https://github.com/Expensify/App/issues/8349#)
